viam-cartographer/src/carto_facade/carto_facade.cc

```diff
@@ -46,38 +86,459 @@ config from_viam_carto_config(viam_carto_config vcc) {
     c.lidar_config = vcc.lidar_config;
     if (c.sensors.size() == 0) {
         throw VIAM_CARTO_SENSORS_LIST_EMPTY;
     }
     if (c.data_dir.size() == 0) {
         throw VIAM_CARTO_DATA_DIR_NOT_PROVIDED;
     }
-    if (c.map_rate_sec < 0) {
+    if (vcc.map_rate_sec < 0) {
         throw VIAM_CARTO_MAP_RATE_SEC_INVALID;
     }
     if (c.component_reference.empty()) {
         throw VIAM_CARTO_COMPONENT_REFERENCE_INVALID;
     }
-    validate_mode(c.mode);
     validate_lidar_config(c.lidar_config);
 
     return c;
 };
 
+std::string find_lua_files() {
+    auto programLocation = boost::dll::program_location();
+    auto localRelativePathToLuas = programLocation.parent_path().parent_path();
+    localRelativePathToLuas.append("lua_files");
+    auto relativePathToLuas = programLocation.parent_path().parent_path();
+    relativePathToLuas.append("share/cartographer/lua_files");
+    fs::path absolutePathToLuas("/usr/local/share/cartographer/lua_files");
+
+    if (exists(relativePathToLuas)) {
+        VLOG(1) << "Using lua files from relative path "
+                << relativePathToLuas.string();
+        return relativePathToLuas.string();
+    } else if (exists(localRelativePathToLuas)) {
+        VLOG(1) << "Using lua files from local relative path "
+                << localRelativePathToLuas.string();
+        return localRelativePathToLuas.string();
+    } else if (exists(absolutePathToLuas)) {
+        VLOG(1) << "Using lua files from absolute path "
+                << absolutePathToLuas.string();
+        return absolutePathToLuas.string();
+    } else {
+        LOG(ERROR) << "No lua files found, looked in " << relativePathToLuas;
+        LOG(ERROR) << "Use 'make install-lua-files' to install lua files into "
+                      "/usr/local/share";
+        return "";
+    }
+};
+
+const std::string action_mode_lua_config_filename(ActionMode am) {
+    switch (am) {
+        case ActionMode::MAPPING:
+            return configuration_mapping_basename;
+            break;
+        case ActionMode::LOCALIZING:
+            return configuration_localization_basename;
+            break;
+        case ActionMode::UPDATING:
+            return configuration_update_basename;
+            break;
+        default:
+            LOG(ERROR)
+                << "action_mode_lua_config_filename: action mode is invalid";
+            throw VIAM_CARTO_SLAM_MODE_INVALID;
+    }
+};
+
 CartoFacade::CartoFacade(viam_carto_lib *pVCL, const viam_carto_config c,
                          const viam_carto_algo_config ac) {
     lib = pVCL;
     config = from_viam_carto_config(c);
     algo_config = ac;
-    path_to_data = config.data_dir + "/data";
-    // TODO: Change to "/internal_state"
-    path_to_internal_state = config.data_dir + "/map";
+    path_to_internal_state = config.data_dir + "/internal_state";
     b_continue_session = true;
 };
 
+void setup_filesystem(std::string data_dir,
```

Comment:
Currently this only creates the data_dir if it doesn't exist if it is within a directory which does exist. TODO: Check if the go code supports recursively creating subdirectories non of which exist & if so, consider adding support for that.

Reply:
I confirmed that currently, the go code doesn't recursively create subdirectories. I.e. it behaves the same as this code does:
https://github.com/viamrobotics/viam-cartographer/blob/main/config/setup.go#L23